File: elixir/pe022.exs
defmodule ProjectEuler do
  def pe022() do
    {:ok, contents} = File.read("../names.txt")
    names = contents |> String.split(",", trim: true)
    new = for name <- names do
      String.replace(name, ~r/\"/, "", global: true)
    end
    sorted = Enum.sort(new)
    sum_name_scores(sorted, 1, 0)
  end
   
  defp sum_name_scores([head|tail], count, sum) do
    score = compute_score(to_charlist(head), 0, count)
    sum_name_scores(tail, count + 1, sum + score)
  end

  defp sum_name_scores([], _, sum) do
    sum
  end

  defp compute_score([head|tail], score, count) do
    compute_score(tail, score + head - 64,  count)
  end

  defp compute_score([], score, count) do
    score * count
  end
end

IO.puts(ProjectEuler.pe022())

File: haskell/pe022.hs
-- this computes projecteuler.net problem 022

import Data.Char
import Data.List

import System.IO

split :: String -> [String]
split "" = [""]
split (',':cs) = "" : split cs
split (c:cs) = (c:cellCompletion) : otherCells
 where cellCompletion : otherCells = split cs

count_chars :: String -> Int -> Int
count_chars [] ssum = ssum
count_chars (x:xs) ssum | x == '"' = count_chars xs ssum
count_chars (x:xs) ssum = count_chars xs ((ssum - 64) + ord x)

count_words :: [String] -> Int -> Int -> Int
count_words [] index ssum = ssum
count_words (x:xs) index ssum = count_words xs (index + 1) (ssum + ((index + 1) * (count_chars x 0)))

pe022 :: IO ()
pe022 = do
  fileHandle <- openFile "../names.txt" ReadMode
  contents <- hGetContents fileHandle
  let sorted = sort (split contents)
  print (count_words sorted 0 0)
  hClose fileHandle

main = pe022
